src/baseline.rs

```diff
@@ -4,69 +4,91 @@ use anyhow::Context as _;
-    ) -> Vec<String> {
-        // Implicit features from optional dependencies have to be added separately
-        // from regular features: https://github.com/obi1kenobi/cargo-semver-checks/issues/265
-        let mut implicit_features: std::collections::BTreeSet<_> = crate_
-            .dependencies()
-            .iter()
-            .filter_map(|dep| dep.is_optional().then_some(dep.name()))
-            .map(|x| x.to_string())
-            .collect();
-        for feature_defn in crate_.features().values().flatten() {
-            // "If you specify the optional dependency with the dep: prefix anywhere
-            //  in the [features] table, that disables the implicit feature."
-            // https://doc.rust-lang.org/cargo/reference/features.html#optional-dependencies
-            if let Some(optional_dep) = feature_defn.strip_prefix("dep:") {
-                implicit_features.remove(optional_dep);
-            }
-        }
-        let regular_features: std::collections::BTreeSet<_> =
-            crate_.features().keys().cloned().collect();
-        let mut all_crate_features = implicit_features;
-        all_crate_features.extend(regular_features);
-        all_crate_features.into_iter().collect()
-    }
-
-    #[allow(unused_variables)]
-    pub(crate) fn get_all_crate_features_from_manifest(path: &std::path::Path) -> Vec<String> {
-        unimplemented!()
-    }
-}
-
-#[allow(dead_code)]
 #[derive(Debug, Clone)]
 enum CrateSource<'a> {
     Registry { crate_: &'a crates_index::Version },
-    ManifestPath { path: &'a Path, name: String },
+    ManifestPath { manifest: &'a Manifest },
 }
 
 impl<'a> CrateSource<'a> {
-    fn name(&self) -> &str {
+    fn name(&self) -> anyhow::Result<String> {
+        Ok(match self {
+            Self::Registry { crate_ } => crate_.name().to_string(),
+            Self::ManifestPath { manifest } => crate::manifest::get_package_name(manifest)?,
+        })
+    }
+
+    fn version(&self) -> anyhow::Result<String> {
+        Ok(match self {
+            Self::Registry { crate_ } => crate_.version().to_string(),
+            Self::ManifestPath { manifest } => crate::manifest::get_package_version(manifest)?,
+        })
+    }
+
+    fn regular_features(&self) -> Vec<String> {
         match self {
-            Self::Registry { crate_ } => crate_.name(),
-            Self::ManifestPath { .. } => unimplemented!(),
+            Self::Registry { crate_ } => crate_.features().keys().cloned().into_iter().collect(),
+            Self::ManifestPath { manifest } => manifest
+                .parsed
+                .features
+                .keys()
+                .cloned()
+                .into_iter()
+                .collect(),
         }
     }
 
-    fn version(&self) -> &str {
-        match self {
-            Self::Registry { crate_ } => crate_.version(),
-            Self::ManifestPath { .. } => unimplemented!(),
+    fn implicit_features(&self) -> std::collections::BTreeSet<String> {
```

Comment:
Another doc comment here would be good.

Reply:
Done in 941dc0fb034be387c2173280fbf22eb0ec9f1220.